File: algorithms/sorting/SelectionSort.java
class SelectionSort {
    public static int [] field = {7, 3, 4, 5, 3, 8, 1};

    public static void main(String[] args) {
        if(args.length < 1) {
            return;
        }
        
        int count = 0;

        int[] sort_arr = create(0, Integer.parseInt(args[0]));
        int[] back_arr = create(1, Integer.parseInt(args[0]));
        int[] rand_arr = create(2, Integer.parseInt(args[0]));

        count += sort(sort_arr);
        System.out.println("Sorted sort count: " + count);
        count = 0;

        count += sort(back_arr);
        System.out.println("Backwards sort count: " + count);
        count = 0;

        //for(int i = 0; i <= 9999; i++) {
            int[] clone = rand_arr;
            count += sort(clone);
        //}
        System.out.println("Random sort count: " + count);
    }

    private static int sort(int[] f) {
        int min, temp, count = 0;
        for (int i = 0; i < f.length; i++) {
            min = i;
            for (int j = i + 1; j < f.length; j++) {
                if (f[j] < f[min]) {
                    min = j;
                }
            }
            if (min != i) {
                temp = f[i];
                f[i] = f[min];
                f[min] = temp;
                count++;
            }
        }
        //System.out.println("Number of switches: " + count);
        return count;
    }

    private static int[] create(int sMethod, int size) {
        int [] f = new int[size];
        switch (sMethod) {
            case 0: for (int i = 0; i < size; i++) {
                    f[i] = i;
                }; return f;
            case 1: for (int i = 0; i < size; i++) {
                    f[i] = size - i;
                }; return f;
            default: for (int i = 0; i < size; i++) {
                    f[i] = (int)(size*Math.random());
                }; return f;
        }
    }

    private static void out(int[] f) {
        for (int i = 0; i < f.length; i++) {
            System.out.print(f[i] + " ");
        }
        System.out.println();
    }
}

// Complexity: O(n^2)
// 
// Best case (everything already sorted): 0 actions taken
// Worst case (not a single number in the correct positions): size/2 actions taken
// Average case (random): sum(i=0; n)

File: algorithms/sorting/selectionsort.cpp
#include <iostream>
#include <vector>
#include <random>

int sort(std::vector<int> f) {
	int min, temp, count = 0;
	for (int i = 0; i < f.size(); i++) {
		min = i;
		for (int j = i + 1; j < f.size(); j++) {
			if (f[j] < f[min]) {
				min = j;
			}
		}
		if (min != i) {
			temp = f[i];
			f[i] = f[min];
			f[min] = temp;
			count++;
		}
	}
	return count;
}

std::vector<int> create(int s_method, int size) {
	std::vector<int> f (size);
	switch (s_method) {
		case 0:
			for (int i = 0; i < size; i++) {
				f[i] = i;	
			}
			break;
		case 1:
			for (int i = 0; i < size; i++) {
				f[i] = size - i;
			}
			break;
		default:
			for (int i = 0; i < size; i++) {
				std::random_device rd;
				std::mt19937 rng(rd());
				std::uniform_int_distribution<int> uni(0, 10);
				f[i] = uni(rng);
			}
			break;
	}
	return f;
}

int main(int argc, char **argv) {
	/*std::vector<int> field = create(2, 10);
	for (int i = 0; i < field.size(); i++) {
		std::cout << field[i] << std::endl;
	}
	return 0;*/

	int best = 0, worst = 0, avg = 0;
	for (int  i = 0; i < 200; i++) {
		std::vector<int> f_best = create(0, 10);
		std::vector<int> f_avg = create(1, 10);
		std::vector<int> f_worst = create(2, 10);

		best += sort(f_best);
		worst += sort(f_worst);
		avg += sort(f_avg);
	}

	std::cout << "Best best: " << best/200 << std::endl;
	std::cout << "Best avg.: " << avg/200 << std::endl;
	std::cout << "Best worst: " << worst/200 << std::endl;
	return 0;
}
